can/io/mf4.py

```diff
@@ -0,0 +1,392 @@
+"""
+Contains handling of MF4 logging files.
+
+MF4 files represent Measurement Data Format (MDF) version 4 as specified by
+the ASAM MDF standard (see https://www.asam.net/standards/detail/mdf/)
+"""
+
+from datetime import datetime
+from pathlib import Path
+import logging
+
+from ..message import Message
+from ..listener import Listener
+from ..util import channel2int
+from .generic import BaseIOHandler
+
+try:
+    from asammdf import MDF4, Signal
+    import numpy as np
+
+    ASAMMDF_AVAILABLE = True
+
+except ImportError as error:
+    ASAMMDF_AVAILABLE = False
+    raise error
+
+
+CAN_MSG_EXT = 0x80000000
+CAN_ID_MASK = 0x1FFFFFFF
+
+STD_DTYPE = np.dtype(
+    [
+        ("CAN_DataFrame.BusChannel", "<u1"),
+        ("CAN_DataFrame.ID", "<u4"),
+        ("CAN_DataFrame.IDE", "<u1"),
+        ("CAN_DataFrame.DLC", "<u1"),
+        ("CAN_DataFrame.DataLength", "<u1"),
+        ("CAN_DataFrame.DataBytes", "(64,)u1"),
+        ("CAN_DataFrame.Dir", "<u1"),
+        ("CAN_DataFrame.EDL", "<u1"),
+        ("CAN_DataFrame.BRS", "<u1"),
+        ("CAN_DataFrame.ESI", "<u1"),
+    ]
+)
+
+ERR_DTYPE = np.dtype(
+    [
+        ("CAN_ErrorFrame.BusChannel", "<u1"),
+        ("CAN_ErrorFrame.ID", "<u4"),
+        ("CAN_ErrorFrame.IDE", "<u1"),
+        ("CAN_ErrorFrame.DLC", "<u1"),
+        ("CAN_ErrorFrame.DataLength", "<u1"),
+        ("CAN_ErrorFrame.DataBytes", "(64,)u1"),
+        ("CAN_ErrorFrame.Dir", "<u1"),
+        ("CAN_ErrorFrame.EDL", "<u1"),
+        ("CAN_ErrorFrame.BRS", "<u1"),
+        ("CAN_ErrorFrame.ESI", "<u1"),
+    ]
+)
+
+RTR_DTYPE = np.dtype(
+    [
+        ("CAN_DataFrame.BusChannel", "<u1"),
+        ("CAN_DataFrame.ID", "<u4"),
+        ("CAN_DataFrame.IDE", "<u1"),
+        ("CAN_DataFrame.DLC", "<u1"),
+        ("CAN_DataFrame.Dir", "<u1"),
+    ]
+)
+
+FD_LEN2DLC = {12: 9, 16: 10, 20: 11, 24: 12, 32: 13, 48: 14, 64: 15}
+
+FD_DLC2LEN = {value: key for key, value in FD_LEN2DLC.items()}
+
+logger = logging.getLogger("can.io.mf4")
+
+
+class MF4Writer(BaseIOHandler, Listener):
+    """Logs CAN data to an ASAM Measurement Data File v4 (.mf4).
```

Comment:
Is the idea not to support the older MDF file format?

Reply:
The BUS logging support was added as a new feature in the MDF version 4.0.0